app/models/concerns/doi_generator.rb

```diff
@@ -47,36 +47,40 @@ def update_datacite_doi
         response = client.get_doi(doi)
         data = JSON[response.body]["data"]
         identifier = data["id"]
-        new_state = visibility_to_state
-        current_state = data["attributes"]["state"]
-        # if state changes from findable to anything else, we hide it
-        if current_state == "findable" && new_state != "findable"
-          json = datacite_api_json(hide=true)
-        else
-          json = datacite_api_json
-        end
+        doi_state = data["attributes"]["state"]
+        action, message = get_action_message(doi_state, self.visibility)
+        json = datacite_api_json(action)
         client.update_metadata(identifier, json)
-        update_doi_metadata_message(current_state, new_state)
+        self.doi_message = message
       rescue RestClient::NotFound
         next
       end
     end
   end
 
-  def update_doi_metadata_message(old_state, new_state)
-    if new_state == 'registered' && old_state != new_state
-      self.doi_message = 'updated_registered'
-    else
-      self.doi_message = 'updated'
-    end
-
+  def get_action_message(current_doi_state, new_state)
+    states = {
+      ["draft", "restricted"] => ["", "draft_restricted"], # do nothing
+      ["draft", "open"] => ["publish", "draft_published"],
+      ["draft", "authenticated"] => ["register", "draft_registered"],
+      ["findable", "restricted"] => ["hide", "hide_findable"],
+      ["findable", "open"] => ["", "already_findable"], # do nothing
+      ["findable", "authenticated"] => ["hide", "hide_findable"],
+      ["registered", "restricted"] => ["", "registered_restricted"], # do nothing
+      ["registered", "open"] => ["publish", "publish_registered"],
+      ["registered", "authenticated"] => ["", "registered_authenticated"] # do nothing
+    }
+    action = states[[current_doi_state, new_state]][0]
+    message = states[[current_doi_state, new_state]][1]
+    return action, message
```

Comment:
minor contraction:

```suggestion
    return states[[current_doi_state, new_state]]
```

Reply:
Huh, I didn't think this would work... 
``` ruby
pry(main)> a
=> ["hello", "world"]
pry(main)> h, w = a
=> ["hello", "world"]
pry(main)> h
=> "hello"
pry(main)> w
=> "world"
```